synapse/storage/databases/main/relations.py

```diff
@@ -295,60 +332,96 @@ def _get_aggregation_groups_for_event_txn(
             "get_aggregation_groups_for_event", _get_aggregation_groups_for_event_txn
         )
 
-    @cached()
-    async def get_applicable_edit(
-        self, event_id: str, room_id: str
-    ) -> Optional[EventBase]:
+    async def _get_applicable_edits(
+        self, event_ids: Iterable[str]
+    ) -> Dict[str, EventBase]:
         """Get the most recent edit (if any) that has happened for the given
-        event.
+        events.
 
         Correctly handles checking whether edits were allowed to happen.
 
         Args:
-            event_id: The original event ID
-            room_id: The original event's room ID
+            event_ids: The original event IDs
 
         Returns:
-            The most recent edit, if any.
+            A map of the most recent edit for each event. A missing event implies
+            there is no edits.
         """
 
+        # A map of the original event IDs to the edit events.
+        edits_by_original = {}
+
+        # Check if an edit for this event is currently cached.
+        event_ids_to_check = []
+        for event_id in event_ids:
+            if event_id not in self.get_applicable_edit:
+                event_ids_to_check.append(event_id)
+            else:
+                edit_event = self.get_applicable_edit[event_id]
+                if edit_event:
+                    edits_by_original[event_id] = edit_event
+
+        # If all events were cached, all done.
+        if not event_ids_to_check:
+            return edits_by_original
+
         # We only allow edits for `m.room.message` events that have the same sender
         # and event type. We can't assert these things during regular event auth so
         # we have to do the checks post hoc.
 
         # Fetches latest edit that has the same type and sender as the
         # original, and is an `m.room.message`.
+        #
+        # TODO Should this ensure it does not return results for state events / redacted events?
         sql = """
-            SELECT edit.event_id FROM events AS edit
+            SELECT original.event_id, edit.event_id FROM events AS edit
```

Comment:
Ideally we would be only pull out the latest edit event per original event, but I haven't found a reasonable way to do that (maybe a lateral join, but that's not supported on sqlite). Any thoughts on how to improve this would be appreciated!

Reply:
Does latest edit mean "edit with largest depth"?

Might we be able to use a trick like https://stackoverflow.com/a/27802817/5252017 ? (Sorry, not an expert here)

Reply:
It means the largest `origin_server_ts` (which is why we're ordering by that descending.)

Reply:
> maybe a lateral join, but that's not supported on sqlite

N.B. SQLite does support lateral joins as long as you don't write the word `LATERAL` — if that's the only thing blocking you, you should be able to work around that by only inserting that word for Postgres.

Reply:
@reivilibre Do you have any reference for that? I've been unable to get it to work (and my searching online has yielded "you can't do this on SQLite").

Reply:
On Postgres you can do `SELECT DISTINCT ON (original.event_id) ...` which will choose the first row (as defined by the `ORDER BY` clause). I don't think SQLite has that, and so you would need to use a different query, at which point you may as well fall back to the old behaviour.

Potentially you could use window functions and `first_value` , which I *think* both postgres and sqlite support, but those sorts of queries really are voodoo magic.

Reply:
❤️ @erikjohnston Thank you! That gave me some breadcrumbs to realize we do something very similar elsewhere:

https://github.com/matrix-org/synapse/blob/3e0536cd2afb5a640619bd872fc27b068ec3eb9b/synapse/storage/databases/main/end_to_end_keys.py#L616-L639

I think we can abstract out the changes in 8400c20086e7747336cc1cc174628ae182788bd7, but that is proving a bit tedious / invasive so I'd like to do it separately.